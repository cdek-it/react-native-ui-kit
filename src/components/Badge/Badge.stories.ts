import type { Meta, StoryObj } from '@storybook/react'

import { Badge } from './Badge'

const meta: Meta<typeof Badge> = {
  title: 'Badge',
  component: Badge,
  args: { severity: 'basic' },
  argTypes: {
    severity: {
      control: 'radio',
      options: ['basic', 'info', 'success', 'warning', 'danger'],
    },
  },
}

export default meta

type Story = StoryObj<typeof Badge>

export const BadgeDot: Story = { args: { dot: true }, argTypes: {} }

export const BadgeNormal: Story = { args: { children: 'Test' }, argTypes: {} }
