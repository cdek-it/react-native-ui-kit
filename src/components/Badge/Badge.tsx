import { memo } from 'react'
import {
  type AccessibilityProps,
  Text,
  View,
  type ViewStyle,
} from 'react-native'
import type { ViewProps } from 'react-native-svg/lib/typescript/fabric/utils'

import { makeStyles } from '../../utils/makeStyles'

export type BadgeSeverity = 'basic' | 'info' | 'success' | 'warning' | 'danger'

interface BadgeBase
  extends AccessibilityProps,
    Pick<ViewProps, 'onLayout' | 'testID'> {
  /**
   * Выбор варианта стиля компонента
   * @default 'basic'
   */
  severity?: BadgeSeverity
  /** Дополнительная стилизация для контейнера компонента */
  style?: ViewStyle
}

interface BadgeText extends BadgeBase {
  /** Текст внутри бейджа **/
  children: string
  /** Отображать бейдж в форме точки **/
  dot?: never
}

interface BadgeDot extends BadgeBase {
  /** Отображать бейдж в форме точки **/
  dot: true
  /** Текст внутри бейджа **/
  children?: never
}

export type BadgeProps = BadgeText | BadgeDot

/**
 * Компонент Badge
 * @param children - Текст внутри бейджа
 * @param dot - Отображать бейдж в форме точки
 * @param severiy - Выбор варианта стиля компонента
 * @param style - Дополнительная стилизация для контейнера компонента
 * @link https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-4871&m=dev
 */
export const Badge = memo<BadgeProps>(
  ({ children, dot, severity = 'basic', style, ...rest }) => {
    const styles = useStyles()

    return (
      <View style={[styles.container, style]} {...rest}>
        {dot ? (
          <View style={[styles.dot, styles[severity]]} />
        ) : (
          <View style={[styles.textBadgeContainer, styles[severity]]}>
            <Text style={styles.textBadge}>{children}</Text>
          </View>
        )}
      </View>
    )
  }
)

const useStyles = makeStyles(({ theme }) => ({
  container: { alignItems: 'flex-start' },
  dot: {
    width: theme.Misc.Badge.badgeDotSize,
    height: theme.Misc.Badge.badgeDotSize,
    borderRadius: 100,
  },
  textBadgeContainer: {
    height: theme.Misc.Badge.badgeHeight,
    paddingHorizontal: theme.Misc.Tag.tagPadding,
    paddingVertical: theme.Misc.Chip.chipPaddingTopBottom,
    justifyContent: 'center',
    borderRadius: 100,
  },
  textBadge: {
    color: theme.Misc.Badge.badgeTextColor,
    fontSize: 10.5,
    fontWeight: 700,
    includeFontPadding: false,
    verticalAlign: 'middle',
  },
  basic: { backgroundColor: theme.Misc.Badge.badgeBg },
  info: { backgroundColor: theme.Button.Severity.Info.Basic.infoButtonBg },
  success: {
    backgroundColor: theme.Button.Severity.Success.Basic.successButtonBg,
  },
  warning: {
    backgroundColor: theme.Button.Severity.Warning.Basic.warningButtonBg,
  },
  danger: {
    backgroundColor: theme.Button.Severity.Danger.Basic.dangerButtonBg,
  },
}))
