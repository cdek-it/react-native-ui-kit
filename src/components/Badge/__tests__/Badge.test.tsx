import { render } from '@testing-library/react-native'

import { Badge, type BadgeProps } from '../Badge'

describe('Badge component tests', () => {
  const snapshotCases: Array<[string, BadgeProps]> = [
    ['dot, severity: basic', { dot: true, severity: 'basic' }],
    ['dot, severity: info', { dot: true, severity: 'info' }],
    ['dot, severity: success', { dot: true, severity: 'success' }],
    ['dot, severity: warning', { dot: true, severity: 'warning' }],
    ['dot, severity: danger', { dot: true, severity: 'danger' }],
    ['severity: basic', { children: '12', severity: 'basic' }],
    ['severity: basic', { children: '12', severity: 'info' }],
    ['severity: basic', { children: '12', severity: 'success' }],
    ['severity: basic', { children: '12', severity: 'warning' }],
    ['severity: basic', { children: '12', severity: 'danger' }],
    ['severity: default', { children: '12' }],
    [
      'with custom style',
      { dot: true, severity: 'basic', style: { margin: 10 } },
    ],
  ]

  test.each(snapshotCases)('%s', (_, props) => {
    const renderedBadge = render(<Badge {...props} />)

    expect(renderedBadge.toJSON()).toMatchSnapshot()
  })
})
