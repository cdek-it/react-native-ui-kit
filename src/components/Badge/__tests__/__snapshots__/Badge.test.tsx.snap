// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component tests dot, severity: basic 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: danger 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#e26e63",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: info 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#49a2ef",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: success 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#68c170",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: warning 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#ffb330",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests severity: basic 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 21,
          "justifyContent": "center",
          "paddingHorizontal": 7,
          "paddingVertical": 0,
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
    >
      12
    </Text>
  </View>
</View>
`;

exports[`Badge component tests severity: basic 2`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 21,
          "justifyContent": "center",
          "paddingHorizontal": 7,
          "paddingVertical": 0,
        },
        {
          "backgroundColor": "#49a2ef",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
    >
      12
    </Text>
  </View>
</View>
`;

exports[`Badge component tests severity: basic 3`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 21,
          "justifyContent": "center",
          "paddingHorizontal": 7,
          "paddingVertical": 0,
        },
        {
          "backgroundColor": "#68c170",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
    >
      12
    </Text>
  </View>
</View>
`;

exports[`Badge component tests severity: basic 4`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 21,
          "justifyContent": "center",
          "paddingHorizontal": 7,
          "paddingVertical": 0,
        },
        {
          "backgroundColor": "#ffb330",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
    >
      12
    </Text>
  </View>
</View>
`;

exports[`Badge component tests severity: basic 5`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 21,
          "justifyContent": "center",
          "paddingHorizontal": 7,
          "paddingVertical": 0,
        },
        {
          "backgroundColor": "#e26e63",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
    >
      12
    </Text>
  </View>
</View>
`;

exports[`Badge component tests severity: default 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 21,
          "justifyContent": "center",
          "paddingHorizontal": 7,
          "paddingVertical": 0,
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#ffffff",
          "fontSize": 10.5,
          "fontWeight": 700,
          "includeFontPadding": false,
          "verticalAlign": "middle",
        }
      }
    >
      12
    </Text>
  </View>
</View>
`;

exports[`Badge component tests with custom style 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      {
        "margin": 10,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  />
</View>
`;
